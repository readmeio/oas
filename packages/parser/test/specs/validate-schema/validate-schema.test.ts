import { describe, it, expect, assert } from 'vitest';

import { SwaggerParser } from '../../../src/index.js';
import * as path from '../../utils/path.js';

describe('Invalid APIs (Swagger 2.0 and OpenAPI 3.x schema validation)', () => {
  it('should pass validation if "options.validate.schema" is false', async () => {
    const api = await SwaggerParser.validate(path.rel('specs/validate-schema/invalid/invalid-response-code.yaml'), {
      validate: { schema: false },
    });
    expect(api).to.be.an('object');
  });

  it('should return all errors', async () => {
    try {
      await SwaggerParser.validate(path.rel('specs/validate-schema/invalid/multiple-invalid-properties.yaml'));
      assert.fail('Validation should have failed, but it succeeded!');
    } catch (err) {
      expect(err).to.be.an.instanceOf(SyntaxError);
      expect(err.message).to.match(/^OpenAPI schema validation failed.\n(.*)+/);

      expect(err.details).to.be.an('array').to.have.length(3);
      expect(err.totalErrors).to.equal(2);

      expect(err.message).to.contain("REQUIRED must have required property 'url'");
      expect(err.message).to.contain('url is missing here');
      expect(err.message).to.contain('ADDITIONAL PROPERTY must NOT have additional properties');
      expect(err.message).to.contain('tagss is not expected to be here');
    }
  });

  it.each([
    {
      name: 'unknown JSON Schema format',
      file: 'unknown-format.yaml',
    },
    {
      name: 'Schema with "allOf"',
      file: 'allof.yaml',
    },
  ])('$name', async ({ file }) => {
    const api = await SwaggerParser.validate(path.rel(`specs/validate-schema/valid/${file}`));
    expect(api).to.be.an('object');
  });

  it.each([
    {
      name: 'invalid response code',
      file: 'invalid-response-code.yaml',
    },
    {
      name: 'optional path param',
      file: 'optional-path-param.yaml',
    },
    {
      name: 'non-required path param',
      file: 'non-required-path-param.yaml',
    },
    {
      name: 'invalid schema type',
      file: 'invalid-schema-type.yaml',
    },
    {
      name: 'invalid param type',
      file: 'invalid-param-type.yaml',
    },
    {
      name: 'non-primitive param type',
      file: 'non-primitive-param-type.yaml',
    },
    {
      name: 'invalid parameter location',
      file: 'invalid-param-location.yaml',
    },
    {
      name: '"file" type used for non-formData param',
      file: 'file-header-param.yaml',
    },
    {
      name: '"file" type used for body param',
      file: 'file-body-param.yaml',
    },
    {
      name: '"multi" header param',
      file: 'multi-header-param.yaml',
    },
    {
      name: '"multi" path param',
      file: 'multi-path-param.yaml',
    },
    {
      name: 'invalid response header type',
      file: 'invalid-response-header-type.yaml',
    },
    {
      name: 'non-primitive response header type',
      file: 'non-primitive-response-header-type.yaml',
    },
    {
      name: 'invalid response schema type',
      file: 'invalid-response-type.yaml',
    },
    {
      name: '$ref to invalid Path object',
      file: 'ref-to-invalid-path.yaml',
    },
    {
      name: 'Schema with "anyOf"',
      file: 'anyof.yaml',
    },
    {
      name: 'Schema with "oneOf"',
      file: 'oneof.yaml',
    },
    {
      name: 'invalid security scheme for OpenAPI 3.0',
      file: 'invalid-security-scheme.yaml',
      isOpenAPI: true,
    },
  ])('$name', async ({ file, isOpenAPI }) => {
    try {
      await SwaggerParser.validate(path.rel(`specs/validate-schema/invalid/${file}`));
      assert.fail('Validation should have failed, but it succeeded!');
    } catch (err) {
      expect(err).to.be.an.instanceOf(SyntaxError);
      if (isOpenAPI) {
        expect(err.message).to.match(/^OpenAPI schema validation failed.\n(.*)+/);
      } else {
        expect(err.message).to.match(/^Swagger schema validation failed.\n(.*)+/);
      }

      expect(err.details).to.be.an('array').with.length.above(0);
      expect(err.totalErrors).to.be.at.least(1);

      // Make sure the Ajv error details object is valid
      const details = err.details[0];
      expect(details.instancePath)
        .to.be.a('string')
        .and.match(/[a-zA-Z/~01]+/); // /paths/~1users/get/responses
      expect(details.schemaPath)
        .to.be.a('string')
        .and.match(/^#\/[a-zA-Z\\/]+/); // #/properties/parameters/items/oneOf
      expect(details.keyword).to.be.a('string').and.match(/\w+/); // oneOf
      expect(details.params).to.be.a('object'); // { passingSchemas: null }
      expect(details.message).to.be.a('string').with.length.of.at.least(1); // must match exactly one schema in oneOf
    }
  });
});
