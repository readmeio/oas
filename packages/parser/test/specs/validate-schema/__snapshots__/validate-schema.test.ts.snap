// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Invalid APIs (Swagger 2.0 and OpenAPI 3.x schema validation) > '"file" type used for body param' 1`] = `
[
  {
    "message": "ENUM must be equal to one of the allowed values
(array, boolean, integer, null, number, object, string)

  18 |           "in": "body",
  19 |           "schema": {
> 20 |             "type": "file"
     |                     ^^^^^^ Did you mean null here?
  21 |           }
  22 |         }
  23 |       ],",
  },
]
`;

exports[`Invalid APIs (Swagger 2.0 and OpenAPI 3.x schema validation) > '"file" type used for non-formData parâ€¦' 1`] = `
[
  {
    "message": "REQUIRED must have required property 'schema'

  14 |           "required": true
  15 |         },
> 16 |         {
     |         ^ schema is missing here!
  17 |           "name": "image",
  18 |           "in": "header",
  19 |           "type": "file"",
  },
]
`;

exports[`Invalid APIs (Swagger 2.0 and OpenAPI 3.x schema validation) > '"multi" header param' 1`] = `
[
  {
    "message": "REQUIRED must have required property 'schema'

   8 |     "/users": {
   9 |       "parameters": [
> 10 |         {
     |         ^ schema is missing here!
  11 |           "name": "usernames",
  12 |           "in": "header",
  13 |           "type": "array",",
  },
]
`;

exports[`Invalid APIs (Swagger 2.0 and OpenAPI 3.x schema validation) > '"multi" path param' 1`] = `
[
  {
    "message": "REQUIRED must have required property 'schema'

   8 |     "/users/{usernames}": {
   9 |       "parameters": [
> 10 |         {
     |         ^ schema is missing here!
  11 |           "name": "usernames",
  12 |           "in": "path",
  13 |           "required": true,",
  },
]
`;

exports[`Invalid APIs (Swagger 2.0 and OpenAPI 3.x schema validation) > '$ref to invalid Path object' 1`] = `
[
  {
    "message": "ADDITIONAL PROPERTY must NOT have additional properties

  16 |     },
  17 |     "/products": {
> 18 |       "type": "object",
     |       ^^^^^^ type is not expected to be here!
  19 |       "properties": {
  20 |         "id": {
  21 |           "type": "number"",
  },
]
`;

exports[`Invalid APIs (Swagger 2.0 and OpenAPI 3.x schema validation) > 'Schema with "anyOf"' 1`] = `
[
  {
    "message": "ADDITIONAL PROPERTY must NOT have additional properties

  12 |             "description": "hello world",
  13 |             "schema": {
> 14 |               "anyOf": [
     |               ^^^^^^^ anyOf is not expected to be here!
  15 |                 {
  16 |                   "properties": {
  17 |                     "firstName": {",
  },
]
`;

exports[`Invalid APIs (Swagger 2.0 and OpenAPI 3.x schema validation) > 'Schema with "oneOf"' 1`] = `
[
  {
    "message": "ADDITIONAL PROPERTY must NOT have additional properties

  12 |             "description": "hello world",
  13 |             "schema": {
> 14 |               "oneOf": [
     |               ^^^^^^^ oneOf is not expected to be here!
  15 |                 {
  16 |                   "properties": {
  17 |                     "firstName": {",
  },
]
`;

exports[`Invalid APIs (Swagger 2.0 and OpenAPI 3.x schema validation) > 'invalid param type' 1`] = `
[
  {
    "message": "REQUIRED must have required property 'schema'

   8 |     "/users": {
   9 |       "parameters": [
> 10 |         {
     |         ^ schema is missing here!
  11 |           "name": "user",
  12 |           "in": "query",
  13 |           "type": "user"",
  },
]
`;

exports[`Invalid APIs (Swagger 2.0 and OpenAPI 3.x schema validation) > 'invalid parameter location' 1`] = `
[
  {
    "message": "REQUIRED must have required property 'schema'

  14 |           "type": "string"
  15 |         },
> 16 |         {
     |         ^ schema is missing here!
  17 |           "name": "img_id",
  18 |           "in": "cookie",
  19 |           "type": "number"",
  },
]
`;

exports[`Invalid APIs (Swagger 2.0 and OpenAPI 3.x schema validation) > 'invalid response code' 1`] = `
[
  {
    "message": "ADDITIONAL PROPERTY must NOT have additional properties

   9 |       "get": {
  10 |         "responses": {
> 11 |           "foobar": {
     |           ^^^^^^^^ foobar is not expected to be here!
  12 |             "description": "hello world"
  13 |           }
  14 |         }",
  },
]
`;

exports[`Invalid APIs (Swagger 2.0 and OpenAPI 3.x schema validation) > 'invalid response header type' 1`] = `
[
  {
    "message": "ENUM must be equal to one of the allowed values
(string, number, integer, boolean, array)

  16 |               },
  17 |               "Last-Modified": {
> 18 |                 "type": "date"
     |                         ^^^^^^ Did you mean number here?
  19 |               }
  20 |             }
  21 |           }",
  },
]
`;

exports[`Invalid APIs (Swagger 2.0 and OpenAPI 3.x schema validation) > 'invalid response schema type' 1`] = `
[
  {
    "message": "ENUM must be equal to one of the allowed values
(array, boolean, integer, null, number, object, string)

  12 |             "description": "hello world",
  13 |             "schema": {
> 14 |               "type": "user"
     |                       ^^^^^^ Did you mean number here?
  15 |             }
  16 |           }
  17 |         }",
  },
]
`;

exports[`Invalid APIs (Swagger 2.0 and OpenAPI 3.x schema validation) > 'invalid schema type' 1`] = `
[
  {
    "message": "ENUM must be equal to one of the allowed values
(array, boolean, integer, null, number, object, string)

  12 |           "in": "body",
  13 |           "schema": {
> 14 |             "type": "user"
     |                     ^^^^^^ Did you mean number here?
  15 |           }
  16 |         }
  17 |       ],",
  },
]
`;

exports[`Invalid APIs (Swagger 2.0 and OpenAPI 3.x schema validation) > 'invalid security scheme for OpenAPI 3â€¦' 1`] = `
[
  {
    "message": "REQUIRED must have required property 'name'

  23 |   "components": {
  24 |     "securitySchemes": {
> 25 |       "tlsAuth": {
     |                  ^ name is missing here!
  26 |         "type": "mutualTLS"
  27 |       }
  28 |     }",
  },
]
`;

exports[`Invalid APIs (Swagger 2.0 and OpenAPI 3.x schema validation) > 'non-primitive param type' 1`] = `
[
  {
    "message": "REQUIRED must have required property 'schema'

   8 |     "/users": {
   9 |       "parameters": [
> 10 |         {
     |         ^ schema is missing here!
  11 |           "name": "user",
  12 |           "in": "query",
  13 |           "type": "object"",
  },
]
`;

exports[`Invalid APIs (Swagger 2.0 and OpenAPI 3.x schema validation) > 'non-primitive response header type' 1`] = `
[
  {
    "message": "ENUM must be equal to one of the allowed values
(string, number, integer, boolean, array)

  16 |               },
  17 |               "Last-Modified": {
> 18 |                 "type": "object"
     |                         ^^^^^^^^ Did you mean boolean here?
  19 |               }
  20 |             }
  21 |           }",
  },
]
`;

exports[`Invalid APIs (Swagger 2.0 and OpenAPI 3.x schema validation) > 'non-required path param' 1`] = `
[
  {
    "message": "REQUIRED must have required property 'schema'

  14 |           "type": "string"
  15 |         },
> 16 |         {
     |         ^ schema is missing here!
  17 |           "name": "img_id",
  18 |           "in": "path",
  19 |           "type": "number"",
  },
]
`;

exports[`Invalid APIs (Swagger 2.0 and OpenAPI 3.x schema validation) > 'optional path param' 1`] = `
[
  {
    "message": "REQUIRED must have required property 'schema'

  14 |           "type": "string"
  15 |         },
> 16 |         {
     |         ^ schema is missing here!
  17 |           "name": "img_id",
  18 |           "in": "path",
  19 |           "required": false,",
  },
]
`;
